src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+
+        // This function loops, because in the case of a blocking read, we may need to
+        // read multiple sets of bytes from the pipe to receive a complete message.
+        unsafe {
+            let mut reader = self.reader.borrow_mut();
+            loop {
+                // First, try to fetch a message, in case we have one in the pipe
+                if let Ok(msg) = reader.get_message() {
+                    return Ok(msg);
+                }
+
+                // If the pipe was already closed, we're done -- we've
+                // already drained all incoming bytes
+                if reader.closed {
+                    return Err(WinError::ok_channel_closed());
+                }
+
+                // Then, issue a read if we don't have one already in flight.
+                // We must not issue a read if we have unconsumed messages,
+                // because getting a message modifies the read_buf.
+                //
+                // XXX is there some rust-y safe way to do this?  I'd like to
+                // transfer ownership of the read buffer to something that can
+                // either consume a message, from the buffer, or have it in flight,
+                // but not both.
+                try!(reader.start_read());
+
+                // If the last read flagged us closed we're done; we've already
+                // drained all incoming bytes earlier in the loop.
+                if reader.closed {
+                    return Err(WinError::ok_channel_closed());
+                }
+
+                // Then, get the overlapped result, blocking if we need to.
+                let mut nbytes: u32 = 0;
+                let mut err = winapi::ERROR_SUCCESS;
+                let ok = kernel32::GetOverlappedResult(reader.handle, reader.ov_ptr(), &mut nbytes,
+                                                       if block { winapi::TRUE } else { winapi::FALSE });
+                if ok == winapi::FALSE {
+                    err = GetLastError();
+                    if !block && err == winapi::ERROR_IO_INCOMPLETE {
+                        // Nonblocking read, no message, read's in flight, we're
+                        // done.
+                        return Err(WinError::ok_no_data());
+                    }
+                    // We pass err through to notify_completion so
+                    // that it can handle other errors.
+                }
+
+                // The IO completed, so the read is no longer in progress
+                reader.read_in_progress = false;
+
+                // Notify that the read completed, which will update the
+                // read pointers
+                try!(reader.notify_completion(err));
+
+                // If we're blocking, keep looping around until we have a complete message.
+                // If we're not blocking, keep looping until GetOverlappedResult gives us
+                // IO_INCOMPLETE.
```

Comment:
... because you can't read in blocking mode if you are using an async/overlapped-io handle :)

I'm not sure what change you'd like to see here -- you mean if we're not blocking we should keep reading until we get a complete message?  I can, my thinking was more that if you're using `try_recv()`, you're prepared for the case of no data being available, and you're not explicitly expecting any data (or you'd `recv`).  If the sending end stalls for whatever reason, I thought it would be better to unblock the receiver here, and finish the message the next time they `try_recv` again.  I don't have a strong position on this though, if you'd prefer the slightly-more-blocky behaviour I'll change it.